discretize/utils/codeutils.py

```diff
@@ -1,16 +1,16 @@
 from __future__ import print_function, division
 import numpy as np
 import sys
-if sys.version_info < (3,):
-    scalarTypes = [float, int, long, np.float_, np.int_]
-else:
-    scalarTypes = [float, int, np.float_, np.int_]
 
 
+scalarTypes = (complex, float, int, np.number)
+if sys.version_info < (3,):
+    scalarTypes += (long, )
+
 def isScalar(f):
-    if type(f) in scalarTypes:
+    if isinstance(f, scalarTypes):
         return True
-    elif isinstance(f, np.ndarray) and f.size == 1 and type(f[0]) in scalarTypes:
+    elif isinstance(f, np.ndarray) and f.size == 1 and isinstance(f[0], scalarTypes):
         return True
     return False
 
```

Comment:
Sorry for commenting on an old PR @lheagy. I just stumbled upon that and I was wondering if there is any benefit of `discretize.utils.codeutils.isScalar(f)` (lines 6-15) over simply `np.size(f) == 1`? Just thought because you extended it for complex number, so it is fragile on the defined `scalarTypes` and will fail for all others (e.g., `int8`, `float32`, `complex64`, etc).

Reply:
No worries @prisae, thanks for raising this. In the `scalarTypes` we do have `np.number`  which `np.int8` and others inherit.

![image](https://user-images.githubusercontent.com/6361812/81618950-b3951400-93a5-11ea-9898-c6f85dc7e6e3.png)

I think that just checking size isn't quite enough. For example, a 2D 1-element array currently returns `False` from isScalar as the first entry is an array. I _think_ that that logic is what we want. 

![image](https://user-images.githubusercontent.com/6361812/81619392-ba705680-93a6-11ea-9f94-0ad1f8ddfce2.png)

For example in the `TensorType` code, we should never be working with 2D arrays. 

https://github.com/simpeg/discretize/blob/master/discretize/utils/matutils.py#L299-L309

However, this might take a bit more thought. Maybe we do want anything that has one element to evaluate `True`?

Reply:
Yes, it might be good to think about it and agree on that, and then write a little docstring with it. Because the behaviour of `isScalar` is not quite obvious from just looking at it.

Here some values

|                  | np.isscalar(x) | np.ndim(x) | np.size(x)==1 | isScalar(x) |
| ---------------- |:--------------:|:----------:|:-------------:|:-----------:|
| 0                | True           |     0      | True          | True        |
| np.array(0)      | False          |     0      | True          | IndexError  |
| np.atleast_1d(0) | False          |     1      | True          | True        |
| np.atleast_2d(0) | False          |     2      | True          | False       |
| np.atleast_3d(0) | False          |     3      | True          | False       |

`isScalar` currently fails for a 0D numpy array. However, instead of fixing `isScalar` for that case we might consider its purpose first.

What is the use-case that a one-element 1D array passes as scalar, but a one-element 2D or 3D array not?

Reply:
Here the failure for the 0D numpy array:
```
In [58]: isScalar(np.array(0))                                                  
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-58-525f9e80cd80> in <module>
----> 1 isScalar(np.array(0))

~/anaconda3/lib/python3.7/site-packages/discretize/utils/codeutils.py in isScalar(f)
     11     if isinstance(f, scalarTypes):
     12         return True
---> 13     elif isinstance(f, np.ndarray) and f.size == 1 and isinstance(f[0], scalarTypes):
     14         return True
     15     return False

IndexError: too many indices for array
```